### DISCUSSION TITLE
Use working-directory for entire job

### DISCUSSION THREAD

**Discussion by dakom** [community/community#25742] [Actions]
<p>Is it possible to specify a <code>working-directory</code> for an entire job? e.g. the equivilent of specifying the same one on every step?</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>When you use <strong>jobs.&lt;job_id&gt;.steps.run</strong> , using the working-directory keyword, you can specify the working directory of where to run the command. You need to add work-directory keyword to all <strong>jobs.&lt;job_id&gt;.steps.run.</strong></p>
<p>Please refer to this document for more details.</p>
<p><a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fhelp.github.com%2Fen%2Factions%2Fautomating-your-workflow-with-github-actions%2Fworkflow-syntax-for-github-actions%23jobsjob_idstepsrun&amp;data=02%7C01%7Cv-yjingz%40microsoft.com%7C32cb918912264c15571c08d7697744eb%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637093833283766979&amp;sdata=QP7AFxAtXf78O7uw%2FoKkalMwITQoKXG%2FC0PoUtUciUM%3D&amp;reserved=0" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idstepsrun</a></p>
<p>If you want to use working-directory for entire job, I would suggest to set a job level environment variable, set its value with the directory, and then use context with ${{ env.variablename}} as the value of work-directory keyword in <strong>jobs.&lt;job_id&gt;.steps.run</strong>.  </p>
<p>For Contexts syntax for GitHub Actions , please refer to:</p>
<p><a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fhelp.github.com%2Fen%2Factions%2Fautomating-your-workflow-with-github-actions%2Fcontexts-and-expression-syntax-for-github-actions%23contexts&amp;data=02%7C01%7Cv-yjingz%40microsoft.com%7C32cb918912264c15571c08d7697744eb%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637093833283766979&amp;sdata=GdEznfPRfaM0FgrgX%2BFjir%2FkQkciQC6qTR0E1PCHe7g%3D&amp;reserved=0" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/contexts-and-expression-syntax-for-github-actions#contexts</a></p>
<p>Please see my example, my repo file structure is like this:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/35b1b55bcb9c2c10a274796fd839862db17af0fc.png" data-download-href="/uploads/short-url/7F02L1ZA7jsGHQRKsJWNSE37ymE.png?dl=1" title="repo dir.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181092966-9143c474-93f6-4f0b-8143-672e84604d15.png" alt="repo dir.png" data-base62-sha1="7F02L1ZA7jsGHQRKsJWNSE37ymE" width="690" height="176" srcset="https://aws1.discourse-cdn.com/github/optimized/1X/35b1b55bcb9c2c10a274796fd839862db17af0fc_2_690x176.png, https://aws1.discourse-cdn.com/github/original/1X/35b1b55bcb9c2c10a274796fd839862db17af0fc.png 1.5x, https://aws1.discourse-cdn.com/github/original/1X/35b1b55bcb9c2c10a274796fd839862db17af0fc.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/35b1b55bcb9c2c10a274796fd839862db17af0fc_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">repo dir.png</span><span class="informations">1022×262 18.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>I add a job level env named “working-directory”. Then set ${{env.working-directory}} as the value of working-directory keyword in <strong>jobs.&lt;job_id&gt;.steps.run</strong>.  </p>
<p>  <img src="https://user-images.githubusercontent.com/24806478/181092967-f10f5904-f1e4-4779-b91a-7920d907c4ba.png" alt="dir.png" data-base62-sha1="16a89zk9MlPA4EyHuPd1B2umDbO" width="401" height="396"></p>

**Reply by rrousselGit** [reply]
Adding to that, you can use `defaults` to set a default `working-directory` instead of changing it everywhere:

```yaml
jobs:
  build:
    runs-on: ubuntu-latest

    defaults:
      run:
        working-directory: ${{ env.working-directory }}

    steps:
      - run: pwd # will respect working-directory
```

**Reply by seiwonpark** [reply]
@Yanjingzhu @rrousselGit Thank you both!

**Reply by dnut** [reply]
This does not help for steps that do not accept the "working-directory" parameter.

**Reply by seiwonpark** [reply]
@dnut Please check out [my code](https://github.com/kookmin-sw/capstone-2023-40/blob/develop/.github/workflows/gradle.yml#L19). If you set command(like in our case `run`), whenever you use `run` steps then it would set `working-directory`.

**Reply by dnut** [reply]
@SeiwonPark The same applies with your code.

**Reply by seiwonpark** [reply]
> @SeiwonPark The same applies with your code.

Oh you meant like permanent checkout not putting the command to all steps.

**Reply by Adamou02** [reply]
@dnut Do you have a solution for "uses" command ?

**Reply by james-wangx** [reply]
You should look `uses` docs.
For example, in "uses" [setup-node](https://github.com/actions/setup-node). We should use `cache-dependency-path` to specify the path to a dependency file:
```yml
steps:
- uses: actions/checkout@v4
- uses: actions/setup-node@v4
  with:
    node-version: 16
    cache: 'npm'
    cache-dependency-path: subdir/package-lock.json
- run: npm ci
- run: npm test
```

**Reply by kubukoz** [reply]
This is an "answer"? It's a workaround :/

**Reply by will-molloy** [reply]
> Adding to that, you can use defaults to set a default working-directory instead of changing it everywhere:

Why is the correct answer hidden, lol

**Comment by bukowa** [comment]
<p>Its a workaround, but a lot of reptition code and not error prone!</p>

**Comment by MatteoGioioso** [comment]
<p>Yes this is not a great solution since you cannot use</p>
<pre><code>uses
</code></pre>
<p>with </p>
<pre><code>working-directory
</code></pre>

**Comment by nodesocket** [comment]
<p>Is there still no way to set a working directory that applies to all the steps in a job?</p>

**Comment by Yanjingzhu** [comment]
<p><a class="mention" href="/nodesocket">@nodesocket</a> I am afraid that I have to say no. What’s your scenario which need to run all steps in a specific path? Let me see whether there’s a work around for you. </p>

**Comment by xsaamiir** [comment]
<p>Hey <a class="mention" href="/yanjingzhu">@yanjingzhu</a> ,</p>
<p>In my case I am using GitHub Actions in a monorepo and I would love to be able the set the working directory on the job level for all steps since I tend to only run a job for a specific sub directory / repo.</p>

**Reply by ktulark** [reply]
Hello, You can use this 
```
jobs:
  build:
    runs-on: ubuntu-latest
    defaults:
      run:
        working-directory: shared-models
```

**Reply by nnnik** [reply]
As established doesn't work for step with uses (aka Actions) therefore a hackfix at best.

**Comment by Yanjingzhu** [comment]
<p>Hi <a class="mention" href="/sharkyze">@sharkyze</a> , </p>
<p>For job level working-directory , I would recommand you to share your scenario and requirement in the <a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" rel="nofollow noopener">Feedback form for GitHub Actions</a>. Thank you for your understanding. </p>

**Comment by ghost** [comment]
<p> <a class="mention" href="/nodesocket">@nodesocket</a>  Doesn’t this get it done: <a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_iddefaultsrun" rel="nofollow noopener">https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_iddefaultsrun</a></p>

**Comment by nodesocket** [comment]
<p><a class="mention" href="/masahigo">@masahigo</a> that worked perfectly. Must have missed that.</p>
<pre><code class="lang-auto">defaults:  
&amp;nbsp; run:  
&amp;nbsp; &amp;nbsp; working-directory: ./foo  
</code></pre>

**Comment by fullkomnun** [comment]
<p>Take a look at: <code>https://dev.to/shofol/run-your-github-actions-jobs-from-a-specific-directory-1i9e</code></p>

**Reply by tshakah** [reply]
[https://dev.to/shofol/run-your-github-actions-jobs-from-a-specific-directory-1i9e](https://dev.to/shofol/run-your-github-actions-jobs-from-a-specific-directory-1i9e)

**Comment by kazuooooo** [comment]
<pre><code class="lang-json">defaults:
  run:
    working-directory: ./your_working_dir
jobs:
  ...
</code></pre>
<p>works for me.</p>

**Reply by janosh** [reply]
Note that this applies only to steps of type `run` , _not_ all steps. If you have a step like

```yml
- name: Set up node
  uses: actions/setup-node@v3
  with:
    cache-dependency-path: <some-npm/yarn/pnpm-lock-file>
```

the `cache-dependency-path` is still relative to repo root, not `working-directory`.

**Comment by davidcorbin-atmosera** [comment]
<p>Consider a trivial case. One Repo has multiple projects each in their own sub-folder. The commands require that the project folder be the current folder…  Now create workflow/actions so that everything happens for a given project.</p>

**Comment by enthal** [comment]
<p>That’s for all <em>jobs</em>, not all <em>tasks</em> in a given job.</p>

**Comment by james-green-affinity** [comment]
<p>Have you found a way to use</p>
<pre><code class="lang-auto">uses
</code></pre>
<p>with</p>
<pre><code class="lang-auto">working-directory
</code></pre>
<p>?</p>

**Comment by roberthstrand** [comment]
<p>You can set working directory as a default for all jobs, default for all steps in a job, or on a per job basis. See the third example <a href="https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idstepsrun" rel="noopener nofollow ugc">here</a>. Does this help you, or did I misunderstand what you are trying to do?</p>

**Reply by uuf6429** [reply]
If you read the previous messages, the `working-directory` setting only applies to `run`... actions will still be applied to the workspace directory.

You can infer this by the fact that it is under 'run':
```yaml
jobs:
  job1:
    runs-on: ubuntu-latest
    defaults:
      run:     <--------
        working-directory: scripts
```
see also: https://github.com/orgs/community/discussions/25742#discussioncomment-4186461

**Comment by itspngu** [comment]
<pre><code class="lang-yaml">jobs:
  job1:
    runs-on: ubuntu-latest
    defaults:
      run:
        working-directory: scripts
</code></pre>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_iddefaults">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">

      <a href="https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_iddefaults" target="_blank" rel="noopener nofollow ugc">GitHub Docs</a>
  </header>

  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/34752364/181092973-a76d958e-c003-44f9-afbb-973f0981ef43.png" class="thumbnail onebox-avatar" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png">

<h3><a href="https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_iddefaults" target="_blank" rel="noopener nofollow ugc">Workflow syntax for GitHub Actions - GitHub Docs</a></h3>

  <p>A workflow is a configurable automated process made up of one or more jobs. You must create a YAML file to define your workflow configuration.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Reply by ornakash** [reply]
It doesn't work for "uses" steps

**Comment by KyleKing** [comment]
I wish there was a `working-dir: inherit`! Something like:

```yaml
name: wd-inherit

on:
  pull_request:

jobs:
  defaults:
    run:
      working-directory: ./proj1
  job1:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v3
    - uses: actions/setup-python@v3
      working-dir: inherit
```

**Reply by Jean1dev** [reply]
![image](https://user-images.githubusercontent.com/26530455/226603875-cdb4641f-c80e-4297-aca4-cc5431322b81.png)

**Comment by skeet70** [comment]
I've upvoted things here to try to increase visibility, but I'd like to elaborate our use case just in case that helps the github team. Similarly to others, we have a monorepo `workflows` containing all the re-usable internal workflows our organization consumes (things like `rust-ci`, `scala-ci`, `deploy-docker`, etc) that our other repos point to and make use of. They pass inputs to tweak the behaviour of the workflow they're using.

To increase org wide stability of CI we want to test those workflows inside the `workflows` monorepo, so if your change to a re-usable workflow is going to break downstream users you find out about it right away on your PR instead of over the next week as CI starts failing across many repos. To accomplish that we've added simple projects to the `workflows` monorepo that reproduce downstream language projects and can exercise the build+test+deploy functionality of the reusable workflows. Each of those projects needs to be in a subfolder though so the re-usable workflow needs to be passed a top level working directory for all its commands to happen in the context of, and some of those re-usable workflows make use of other actions (so `defaults: run: working-directory`) doesn't cover everything.

**Comment by ashishjullia** [comment]
Any solution for this in 2024?

**Comment by jsh95311** [comment]
Any solution for this in 2025?

**Reply by Adamou02** [reply]
Any solution for this in 2026 😂?
